src/shims/unix/mem.rs

```diff
@@ -0,0 +1,240 @@
+        let ptr = this.reallocate_ptr(
+            old_address,
+            Some((Size::from_bytes(old_size), align)),
+            Size::from_bytes(new_size),
+            align,
+            MiriMemoryKind::Mmap.into(),
+        )?;
+        if let Some(increase) = new_size.checked_sub(old_size) {
+            // We just allocated this, the access is definitely in-bounds and fits into our address space.
+            // mmap guarantees new mappings are zero-init.
+            this.write_bytes_ptr(
+                ptr.offset(Size::from_bytes(old_size), this).unwrap().into(),
+                std::iter::repeat(0u8).take(usize::try_from(increase).unwrap()),
+            )
+            .unwrap();
+        }
+        // Memory mappings are always exposed
+        let (prov, _) = ptr.into_parts();
+        let Provenance::Concrete { alloc_id, tag } = prov else { unreachable!() };
+        intptrcast::GlobalStateInner::expose_ptr(this, alloc_id, tag)?;
+
+        Ok(Scalar::from_pointer(ptr, this))
+    }
+
+    fn munmap(
+        &mut self,
+        addr: &OpTy<'tcx, Provenance>,
+        length: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, Scalar<Provenance>> {
+        let this = self.eval_context_mut();
+
+        let addr = this.read_pointer(addr)?;
+        let length = this.read_scalar(length)?.to_machine_usize(this)?;
+
+        // addr must be a multiple of the page size
+        #[allow(clippy::integer_arithmetic)] // PAGE_SIZE is nonzero
+        if addr.addr().bytes() % this.machine.page_size != 0 {
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")))?;
+            return Ok(Scalar::from_i32(-1));
+        }
+
+        // All pages containing a part of the indicated range are unmapped.
+        let length = this.machine.round_up_to_multiple_of_page_size(length).unwrap_or(u64::MAX);
+
+        let unmap_range = addr.addr()..addr.addr() + Size::from_bytes(length);
+
+        // POSIX says that it is unspecified whether the address range needs to be mapped,
+        // Linux is clear that it is not an error to munmap a region which is not mapped.
+        let mapping_exists = this.memory.alloc_map().iter(|it| {
+            for (id, (_kind, alloc)) in it
+                .filter(|(_id, (kind, _alloc))| *kind == MemoryKind::Machine(MiriMemoryKind::Mmap))
+            {
+                let base_addr = intptrcast::GlobalStateInner::rel_ptr_to_addr(
+                    this,
+                    Pointer::new(*id, Size::ZERO),
+                );
+                let base_addr = Size::from_bytes(base_addr);
+                let alloc_range = base_addr..base_addr + alloc.size();
+
+                if alloc_range == unmap_range {
+                    return Ok(true);
+                } else if alloc_range.overlaps(&unmap_range) {
+                    throw_unsup_format!("Miri does not support partial munmap");
+                }
+            }
+            Ok::<bool, InterpError<'_>>(false)
+        })?;
```

Comment:
Uh, this is not nice...

Is it possible to use `get_alloc_info` instead?

Reply:
I agree that this is not nice.

But the `mmap` API is based on addresses or pages, not allocations. And `get_alloc_info` is based on looking up an `AllocId`, which we may not have. So if the interpreter memory is based on a `HashMap<AllocId, Allocation>` we are kind of out of luck here.

Reply:
Ah! Yes. This is the problem I was stuck on.

I can't figure out how to make this actually work on top of the existing memory. The implementation here is all lined up to pull this off, then we have to call `deallocate_ptr` in order to deallocate the memory, which fails if we don't have the right AllocId for the allocation that we matched.

Reply:
> But the mmap API is based on addresses or pages, not allocations.

You can use `Pointer::from_addr` and `ptr_get_alloc_id`.

Reply:
Ah no, `Pointer::from_addr` is the wrong thing (that function should probably be renamed...)

But we have `ptr_from_addr_cast`. So if you truly need to ignore the provenance of the addr input, that is the way to go.

Reply:
> The implementation here is all lined up to pull this off, then we have to call deallocate_ptr in order to deallocate the memory, which fails if we don't have the right AllocId for the allocation that we matched.

Also if provenance truly doesn't matter here there should probably be a test that uses `ptr::invalid` pointers for deallocation?